### PR TITLE
Add Datastoresyncer tests

### PR DESCRIPTION
**Pull request by tpantelis**
Also:
    

- removed unused kubeClientSet field and ctor param
- changed DatastoreSyncer to take a submariner ClusterInterface and EndpointInterface instead of a submarinerClientset.Interface as the former are much easier to mock failures.
- Modified ensureExclusiveEndpoint to return errors on deletion failures. This makes sense sinnce such failures mean exclusivity couldn't be achieved. I think the reason it just logged the errors before was b/c the original author was not aware of how to distinguish the "not found" error in order to ignore it.

### CODE REVIEW

File: pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
...
+	submarinerClusters  *fakeClientsetv1.FailingClusters
+	submarinerEndpoints *fakeClientsetv1.FailingEndpoints
+	stopCh              chan struct{}
+	runCompleted        chan error
+	handledError        chan error
+	savedErrorHandlers  []func(error)
+	initialObjs         []runtime.Object
+}
+
+func newTestDiver() *testDriver {
```
Comment (mangelajo) [inline pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go]: typo
```suggestion
func newTestDriver() *testDriver {
```

File: pkg/controllers/datastoresyncer/datastore_cluster_sync_test.go [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+var _ = Describe("", func() {
+	Context("Local Cluster syncing", testLocalClusterSyncing)
+	Context("Remote Cluster syncing", testRemoteClusterSyncing)
+})
+
+func testLocalClusterSyncing() {
+	var d *testDriver
+
+	BeforeEach(func() {
+		d = newTestDiver()
```
Comment (mangelajo) [inline pkg/controllers/datastoresyncer/datastore_cluster_sync_test.go]:
```suggestion
		d = newTestDriver()
```

File: pkg/controllers/datastoresyncer/datastore_cluster_sync_test.go [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+
+func testRemoteClusterSyncing() {
+	var (
+		d               *testDriver
+		remoteCluster   *types.SubmarinerCluster
+		onClusterChange datastore.OnClusterChange
+	)
+
+	BeforeEach(func() {
+		d = newTestDiver()
```
Comment (mangelajo) [inline pkg/controllers/datastoresyncer/datastore_cluster_sync_test.go]:
```suggestion
		d = newTestDriver()
```

File: pkg/controllers/datastoresyncer/datastore_endpoint_sync_test.go [hunk 4]
```diff
@@ -0,0 +1,307 @@
...
+	Context("Local Endpoint syncing", testLocalEndpointSyncing)
+	Context("Remote Endpoint syncing", testRemoteEndpointSyncing)
+	Context("Failures while ensuring Endpoint exclusivity", testEndpointExclusivityFailures)
+})
+
+func testLocalEndpointSyncing() {
+	var d *testDriver
+
+	BeforeEach(func() {
+		d = newTestDiver()
```
Comment (mangelajo) [inline pkg/controllers/datastoresyncer/datastore_endpoint_sync_test.go]:
```suggestion
		d = newTestDriver()
```

File: pkg/controllers/datastoresyncer/datastore_endpoint_sync_test.go [hunk 5]
```diff
@@ -0,0 +1,307 @@
...
+
+func testRemoteEndpointSyncing() {
+	var (
+		d                *testDriver
+		remoteEndpoint   *types.SubmarinerEndpoint
+		onEndpointChange datastore.OnEndpointChange
+	)
+
+	BeforeEach(func() {
+		d = newTestDiver()
```
Comment (mangelajo) [inline pkg/controllers/datastoresyncer/datastore_endpoint_sync_test.go]:
```suggestion
		d = newTestDriver()
```

File: pkg/controllers/datastoresyncer/datastore_endpoint_sync_test.go [hunk 6]
```diff
@@ -0,0 +1,307 @@
...
+func testEndpointExclusivityFailures() {
+	var (
+		expectedErr      error
+		d                *testDriver
+		existingEndpoint *types.SubmarinerEndpoint
+	)
+
+	BeforeEach(func() {
+		expectedErr = nil
+		d = newTestDiver()
```
Comment (mangelajo) [inline pkg/controllers/datastoresyncer/datastore_endpoint_sync_test.go]:
```suggestion
		d = newTestDriver()
```

**Comment by tpantelis** [comment]
> a nit...

I'd say a misspelling is more than a nit 😄

File: pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go (hunk 1 above)
Comment (tpantelis) [reply pkg/controllers/datastoresyncer/datastoresyncer_suite_test.go]: Fixed

**Pull request closed without merging**